Improve Bogus carrier and limit stub usage

fredericboivin opened the pull request:
Current spec is using fake classes but stubs every functions call and responses. This PR fixes the Bogus carrier / Bogus calculator so they can be used in the same manner as the live carriers / calculators

Instead of stubbing every function calls, we're just stubbing the responses when we need to raise errors or return empty/wrong response

Commit:
Tweak the bogus classes to they can used without stubbing everything

Commit:
Limit stub usage in active_shipping_spec

Instead of stubbing every function call, we use a live (bogus)
carrier and stub only the response needed to simulate wrong or
empty responses

fredericboivin commented on spec/lib/spree/active_shipping/bogus_carrier.rb:
```diff
@@ -3,12 +3,15 @@ module ActiveShipping
     # Bogus carrier useful for testing.  For some reasons the plugin version of this class does not work
     # properly (it fails to return a RateResponse)
     class BogusCarrier < ::ActiveShipping::Carrier
+      include RSpec::Mocks::ExampleMethods
```
Not sure about that but it's needed below in find_rates

forkata replied:
We shouldn't need to do this. It seems like we have two versions of rspec-mocks, and one is locked because of this https://github.com/solidusio-contrib/solidus_active_shipping/blob/master/solidus_active_shipping.gemspec#L25
Can you try removing that from the gemspec since we should not need it, removing the `Gemfile.lock` and bundling.

fredericboivin replied:
Sure

dangerdogz replied:
I removed the duplicate in the gemspec, but I still need to include Rspec::Mocks because the find_rates function use instance_double to stub the RateResponse from ActiveShipping because it has a billion parameters otherwise

I switched the include for a require spec_helper instead, is it okay?

fredericboivin commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -97,12 +87,8 @@
-
     it 'should use the carrier supplied in the initializer' do
-      expect(carrier).to receive(:find_rates) { response }
+      expect(calculator.carrier).to be_an_instance_of(carrier.class)
```
Edited this test because I had a hard time obtaining the same carrier has the one in the calculator since it's always a new instance, besides we only need to know the class, doesn't matter which instance it is

forkata replied:
I don't see any value in this test, I think we can remove it altogether. The original test did nothing anyways, and we know that the carrier will be an instance of the carrier class, so no need for this I think?

fredericboivin replied:
I think so too, I guess maybe you can test that the calculator is using the proper class instead of trying to initialize another carrier but I'm totally fine with removing it

fredericboivin committed:
Improve the caching tests

Old caching tests we're passing but only because the instance
we expect to receive the find_rates call is not the same instance
as the one used by the calculator so the test can generate false
positives

fredericboivin commented:
Also added a commit to make sure that the caching tests are giving the proper results. Without stubbing the carrier method of the calculator, we get different instances between the one returned by the calculator and the one we create in the test file

So if expect(carrier).not_to receive(:something), it works but only because the instance actually receiving (or not receiving) the function call is not the one we expect to (it's returned by the calculator)

fredericboivin commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -120,39 +103,62 @@
+        # pass but only because carrier we're watching and the carrier
+        # used by the calculator are different)
+        allow(calculator).to receive(:carrier).and_return(carrier)
+        allow(carrier).to receive(:find_rates).and_call_original
```
This is needed because since we're stubbing the carrier, the response returned by stubbed find_rates from the instance double is nil so the compute_package call is crashing because it's expecting a response from ActiveShipping

forkata commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -15,15 +15,13 @@
 
   let(:carrier) { Spree::ActiveShipping::BogusCarrier.new }
```
Since we've now fixed the `BogusCalculator` to properly create the `BogusCarrier`, can we remove this and do any stubbing directly on `calculator.carrier`? Then we can also remove all stubbing that returns this instance of the `carrier` on the `calculator`. This is just an idea, not sure if it will work. We may need to memoize the carrier instance in the bogus calculator so it doesn't create a new one every time we reference `calculator.carrier`.

fredericboivin replied:
Good point

fredericboivin replied:
Fixed it, way simpler now. Thanks!

fredericboivin committed:
Remove useless test about carrier in initializer

fredericboivin committed:
Access carrier through the calculator

Remove the initialization and use of carrier, and access it
through the calculator, this simplify the testing and stubbing
in general. Remove the include of Rspec Mock in the Bogus carrier
but require spec_helper instead because we still need to access
instance_double to stub the RateResponse

fredericboivin commented:
Addressed the feedback. rspec-activemodel-mocks is needed in the weight_limits spec so it's out of scope to remove it there after all. 

And the BogusCarrier needs to include RSpec::Mocks::ExampleMethods (unless there's a better way?) because of the usage of instance_double in its find_rates method
